### PR TITLE
[Feature] Feature scaler tool

### PR DESCRIPTION
**Pull request by roya0045**
## Description

I felt like this was missing and I needed to increase my stack of PR to be reviewed.

This allows to scale features while editing. The scaling is done from the center of the feature or the extent of the selected features.

Here is a gif with the wonky rubberband display that I can't fix for now. (Tried too many offset methods and none work).
scaler_v1

### CODE REVIEW

**Comment by roya0045** [comment]
Lesson learned: Use the damn transform matrix, don't chain transform, its not the same and it will alleviate headaches.
scaler_V2

Test is the only thing missing now...

File: src/app/qgsmaptoolscalefeature.h [hunk 1]
```diff
@@ -0,0 +1,125 @@
+/***************************************************************************
+    qgsmaptoolscalefeature.h  -  map tool for scaling features by mouse drag
+    ---------------------
+    begin                :
```
Comment (nyalldawson) [inline src/app/qgsmaptoolscalefeature.h:4]: populate

File: src/app/qgsmaptoolscalefeature.cpp [hunk 2]
```diff
@@ -0,0 +1,479 @@
+﻿/***************************************************************************
+    qgsmaptoolscalefeature.cpp  -  map tool for scaling features by mouse drag
+    ---------------------
+    begin                :
```
Comment (nyalldawson) [inline src/app/qgsmaptoolscalefeature.cpp:4]: populate

File: src/app/qgsmaptoolscalefeature.cpp [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+  mScaleSpinBox->setValue( 1 );
+  mScaleSpinBox->setShowClearButton( false );
+  mScaleSpinBox->setSizePolicy( QSizePolicy::MinimumExpanding, QSizePolicy::Preferred );
+  mLayout->addWidget( mScaleSpinBox );
+
+  mMagnetSpinBox = new QgsSpinBox( this );
+  mMagnetSpinBox->setMinimum( 0 );
+  mMagnetSpinBox->setMaximum( 180 );
+  mMagnetSpinBox->setPrefix( tr( "Snap to " ) );
+  mMagnetSpinBox->setSuffix( tr( "°" ) );
```
Comment (nyalldawson) [inline src/app/qgsmaptoolscalefeature.cpp:66]:
```suggestion
  mMagnetSpinBox->setSuffix( tr( "×" ) );
```

File: src/app/qgsmaptoolscalefeature.cpp [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+  mAnchorPoint.reset();
+  deleteRubberband();
+}
+
+void QgsMapToolScaleFeature::canvasMoveEvent( QgsMapMouseEvent *e )
+{
+  if ( mScalingActive )
+  {
+    const double distance = mFeatureCenter.distance( toLayerCoordinates( mLayer, e->mapPoint() ) );
+    double scale =  distance / mBaseDistance; // min 0 or no limit?
```
Comment (nyalldawson) [inline src/app/qgsmaptoolscalefeature.cpp:155]: Better handle the situation where mBaseDistance == 0

File: src/app/qgsmaptoolscalefeature.cpp [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+
+void QgsMapToolScaleFeature::canvasReleaseEvent( QgsMapMouseEvent *e )
+{
+  if ( !mCanvas )
+  {
+    return;
+  }
+
+  //QgsVectorLayer *mLayer = currentVectorLayer();
+  if ( !mLayer )
```
Comment (nyalldawson) [inline src/app/qgsmaptoolscalefeature.cpp:182]: What's the rationale behind the change to storing the layer instead of using currentVectorLayer()?

File: src/app/qgsmaptoolscalefeature.cpp [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+  }
+
+  // place anchor point on CTRL + click
+  if ( e->modifiers() & Qt::ControlModifier )
+  {
+    if ( !mAnchorPoint )
+    {
+      return;
+    }
+    mFeatureCenter = toLayerCoordinates( mLayer, e->mapPoint() );;
```
Comment (nyalldawson) [inline src/app/qgsmaptoolscalefeature.cpp:203]:
```suggestion
    mFeatureCenter = toLayerCoordinates( mLayer, e->mapPoint() );
```

File: src/app/qgsmaptoolscalefeature.cpp [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+    mAnchorPoint->setCenter( mFeatureCenter );
+    return;
+  }
+
+  deleteScalingWidget();
+
+  // Initialize rotation if not yet active
```
Comment (nyalldawson) [inline src/app/qgsmaptoolscalefeature.cpp:210]:
```suggestion
  // Initialize scaling if not yet active
```

File: src/app/qgsmaptoolscalefeature.cpp [hunk 8]
```diff
@@ -0,0 +1,479 @@
...
+    mRubberBand->show();
+    mBaseDistance = toLayerCoordinates( mLayer, e->mapPoint() ).distance( mFeatureCenter );
+    mScaling = 1.0;
+
+    createScalingWidget();
+    if ( e->modifiers() & Qt::ShiftModifier )
+    {
+      if ( mScalingWidget )
+      {
+        mScalingWidget->setMagnet( 45 );
```
Comment (nyalldawson) [inline src/app/qgsmaptoolscalefeature.cpp:307]: 45 has no meaning here...

File: src/app/qgsmaptoolscalefeature.cpp [hunk 9]
```diff
@@ -0,0 +1,479 @@
...
+  }
+}
+
+
+void QgsMapToolScaleFeature::applyScaling( double scale )
+{
+  mScaling = scale;
+  mScalingActive = false;
+
+  QgsVectorLayer *mLayer = currentVectorLayer();
```
Comment (nyalldawson) [inline src/app/qgsmaptoolscalefeature.cpp:354]:
```suggestion
  QgsVectorLayer *vlayer = currentVectorLayer();
```

Only member variables should use the m prefix

File: src/app/qgsmaptoolscalefeature.cpp [hunk 10]
```diff
@@ -0,0 +1,479 @@
...
+
+  //calculations for affine transformation
+
+  mLayer->beginEditCommand( tr( "Features Scaled" ) );
+
+  int start = ( mLayer->geometryType() == 2 )? 1 : 0;
+
+  int i = 0;
+  const auto constMScaledFeatures = mScaledFeatures;
+  for ( QgsFeatureId id : constMScaledFeatures )
```
Comment (nyalldawson) [inline src/app/qgsmaptoolscalefeature.cpp:370]:
```suggestion
  for ( QgsFeatureId id : qgis::as_const( mScaledFeatures ) )
```

File: src/app/qgsmaptoolscalefeature.cpp [hunk 11]
```diff
@@ -0,0 +1,479 @@
...
+  {
+    QgsFeature feat;
+    mLayer->getFeatures( QgsFeatureRequest().setFilterFid( id ) ).nextFeature( feat );
+    QgsGeometry geom = feat.geometry();
+    i = start;
+    QgsPointXY vertex = geom.vertexAt( i );
+    while ( !vertex.isEmpty() )
```
Comment (nyalldawson) [inline src/app/qgsmaptoolscalefeature.cpp:377]: Much more efficient would be to construct a QTransform and use QgsGeometry::transform here, instead of the vertex by vertex manipulation.

**Review by nyalldawson — COMMENTED** [review]
Looking good, and I agree this is a much wanted feature!

Somethings to consider:
1. I'm not sure how useful the magnet option is with scaling. I'd think about removing that entirely.
2. For maximum flexibility we should offer non proportional scaling where the x/y scaling factors are different, and use the shift modifier to enable aspect ratio constrained scaling.

File: src/app/qgsmaptoolscalefeature.cpp (hunk 5 above)
Comment (roya0045) [reply src/app/qgsmaptoolscalefeature.cpp:182]: Easier retrieval mostly.

**Comment by roya0045** [comment]
> 1. I'm not sure how useful the magnet option is with scaling. I'd think about removing that entirely.
> 2. For maximum flexibility we should offer non proportional scaling where the x/y scaling factors are different, and use the shift modifier to enable aspect ratio constrained scaling.
1. Sure thing then.
2. I'll stick to the base work, if there is a need for such thing follow up PR could address this, I'm fine with this as of now.

Thanks for the review.

File: src/app/qgsmaptoolscalefeature.cpp (hunk 11 above)
Comment (roya0045) [reply src/app/qgsmaptoolscalefeature.cpp:377]: Qtransform seems risky, but using the currend transform could be doable.

File: src/app/qgsmaptoolscalefeature.cpp (hunk 4 above)
Comment (roya0045) [reply src/app/qgsmaptoolscalefeature.cpp:155]: Raise an error or very small offset?

**Comment by roya0045** [comment]
@nyalldawson Looking back I like the magnet widget, just because it displays the scale and helps you know how far you have gone.  This is a nice UX thing.

File: src/app/qgsmaptoolscalefeature.cpp (hunk 9 above)
Comment (roya0045) [reply src/app/qgsmaptoolscalefeature.cpp:354]: removed that assignation, using cached layer.

File: src/app/qgsmaptoolscalefeature.cpp (hunk 8 above)
Comment (roya0045) [reply src/app/qgsmaptoolscalefeature.cpp:307]: removed

File: src/app/qgsmaptoolscalefeature.cpp (hunk 7 above)
Comment (roya0045) [reply src/app/qgsmaptoolscalefeature.cpp:210]: done

File: src/app/qgsmaptoolscalefeature.cpp (hunk 6 above)
Comment (roya0045) [reply src/app/qgsmaptoolscalefeature.cpp:203]: done

File: src/app/qgsmaptoolscalefeature.cpp (hunk 11 above)
Comment (roya0045) [reply src/app/qgsmaptoolscalefeature.cpp:377]: Tried it, get succes but geometry didn't change...unsure why it doesn't work. Will upload other fixes once I figure out why it fails or revert it to the vertex approach.

File: src/app/qgsmaptoolscalefeature.cpp (hunk 3 above)
Comment (roya0045) [reply src/app/qgsmaptoolscalefeature.cpp]: done

File: src/app/qgsmaptoolscalefeature.cpp (hunk 11 above)
Comment (roya0045) [reply src/app/qgsmaptoolscalefeature.cpp:371]: Reverted to vertex, commented out transform approach until it works.

File: src/app/qgsmaptoolscalefeature.cpp (hunk 4 above)
Comment (roya0045) [reply src/app/qgsmaptoolscalefeature.cpp:159]: went for return only.

File: src/app/qgsmaptoolscalefeature.cpp (hunk 2 above)
Comment (roya0045) [reply src/app/qgsmaptoolscalefeature.cpp]: done, didn't put my email though, still not at ease with that...

**Comment by roya0045** [comment]
Did a test and the corner widget is great to set the exact scale instead of fiddling with the mouse.

**Comment by DelazJ** [comment]
Great, @roya0045 
> The scaling is done from the center of the feature or the extent of the selected features.

Maybe, for a follow-up PR, allow to select a custom position as center point of the scaling? Using ctrl+click, you can currently select another center for the rotation tool. Would align, too.

**Comment by roya0045** [comment]
@DelazJ I didn't test that but the code is still there, it should work.

**Comment by nyalldawson** [comment]
@roya0045 while looking into this I ran into https://github.com/qgis/QGIS/pull/40883 -- I'll get that one resolved first since the changes will also need to be applied to this branch.

I gave this a test run though, and it's working nicely! As soon as https://github.com/qgis/QGIS/pull/40883 is merged I'll make a copy of this branch and apply a few minor fix ups, and then we can merge...

**Comment by roya0045** [comment]
@nyalldawson Excellent news! As usual thanks a ton for taking the time to look into my work, this is greatly appreciated.

**Comment by nyalldawson** [comment]
@roya0045 I spotted one issue with this tool - if you start a resize operation and then change the map view (eg zoom in or out) then the rubber band preview goes all wonky.

**Comment by roya0045** [comment]
@nyalldawson can you define what 'all wonky' means? The scale should be handled in map units, so that part should be fine, but the rubber band should be anchored in map units too but the transform to recenter is in screen units since these are the base units for the rubber band.

This might be why things go wonky, if I still have it in my linux vm I'll try to see what's wrong and how it could be solved, the easiest might be the create a new rubber band as the anchor position and scales should be fine.

**Comment by nyalldawson** [comment]
> you define what 'all wonky' means

It gets offset from the actual features

**Comment by roya0045** [comment]
@nyalldawson that's what I was expecting, I assum that the center is the same position in the screen, but not the map. Not sure if its better to create a new one or try to move it properly by refreshing the anchor position or the offsets. At least this isn't suprising.

**Comment by roya0045** [comment]
@nyalldawson Wonkyness has been resolved. Moved the refresh to a distinct function. The new function needs a double since the change scale signal sends out a double.

**Comment by nyalldawson** [comment]
Superseded by https://github.com/qgis/QGIS/pull/40951
